ipam/allocator.go

```diff
@@ -159,97 +187,293 @@ const (
 	v6 = 6
 )
 
-/*******************
- * IPAMConf Contract
- ********************/
+/******************
+ * Config Contract
+ ******************/
+
+// GetDefaultAddressSpaces returns the local and global default address spaces
+func (a *Allocator) GetDefaultAddressSpaces() (string, string, error) {
+	return localAddressSpace, globalAddressSpace, nil
+}
+
+// RequestPool returns an address pool along with its unique id.
+func (a *Allocator) RequestPool(addressSpace, pool, subPool string, options map[string]string, v6 bool) (string, *net.IPNet, map[string]string, error) {
+	var (
+		nw, pw   *net.IPNet
+		ipr      *addressRange
+		c, p, pp *PoolData
+		ok       bool
+		err      error
+	)
 
-// AddSubnet adds a subnet for the specified address space
-func (a *Allocator) AddSubnet(addrSpace AddressSpace, subnetInfo *SubnetInfo) error {
-	// Sanity check
-	if addrSpace == "" {
-		return ErrInvalidAddressSpace
+	if addressSpace == "" {
+		return "", nil, nil, ipamapi.ErrInvalidAddressSpace
 	}
-	if subnetInfo == nil || subnetInfo.Subnet == nil {
-		return ErrInvalidSubnet
+
+	if pool == "" && subPool != "" {
+		return "", nil, nil, ipamapi.ErrInvalidSubPool
 	}
-	// Convert to smaller internal subnets (if needed)
-	subnetList, err := getInternalSubnets(subnetInfo.Subnet, a.internalHostSize)
-	if err != nil {
-		return err
+
+	if pool != "" {
+		_, nw, err = net.ParseCIDR(pool)
+		if err != nil {
+			return "", nil, nil, ipamapi.ErrInvalidPool
+		}
+		if pw, err = adjustAndCheckSubnetSize(nw); err != nil {
+			return "", nil, nil, err
+		}
+		// Reset pool to canonical form
+		pool = types.GetIPNetCanonical(nw).String()
+		if subPool != "" {
+			_, sub, err := net.ParseCIDR(subPool)
+			if err != nil {
```

Comment:
Can this whole code snippet be simplified? Given a `sub` you need to write a function which gives the lower IP and higher IP. I am not sure we need to expose all the low level steps in this function. Also you should be able to get the highest value by ORing the host part. This function is invoking too many functions(`GetHostPartIP/GetBroadcastIP/GetMinimalIP`) to achieve this